src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,109 @@
+        indexOfLastAddedElement++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > indexOfLastAddedElement || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index + ", Size: " + indexOfLastAddedElement);
+        } else {
+            if (indexOfLastAddedElement == array.length) {
+                array = Arrays.copyOf(array, (int) (array.length * 1.5));
+            }
+            Object[] secondPart = Arrays.copyOfRange(array, index, indexOfLastAddedElement);
+            array[index] = value;
+            indexOfLastAddedElement++;
+            for (int i = index + 1, q = 0; i < indexOfLastAddedElement; i++, q++) {
+                array[i] = secondPart[q];
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (indexOfLastAddedElement + list.size() >= array.length) {
+            array = Arrays.copyOf(array, (int) (array.length * 1.5));
+        }
+        for (int i = indexOfLastAddedElement, q = 0; q < list.size(); i++, q++) {
+            array[i] = list.get(q);
+            indexOfLastAddedElement++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (indexOfLastAddedElement <= index || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index + ", Size: " + indexOfLastAddedElement);
+        }
+        return (T) array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (indexOfLastAddedElement <= index || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index + ", Size: " + indexOfLastAddedElement);
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (indexOfLastAddedElement <= index || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index + ", Size: " + indexOfLastAddedElement);
+        }
+        T oldValue = (T) array[index];
+        Object[] secondPart = Arrays.copyOfRange(array, index + 1, indexOfLastAddedElement);
+        indexOfLastAddedElement--;
+        for (int i = index, q = 0; i < indexOfLastAddedElement; i++, q++) {
+            array[i] = secondPart[q];
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T oldValue = null;
+        int counter = 0;
+        for (int i = 0; i < array.length; i++) {
+            if (Objects.equals(array[i], element)) {
+                oldValue = (T) array[i];
+                remove(i);
+                counter++;
+                break;
+            }
+        }
+        if (counter == 0) {
+            throw new NoSuchElementException("no such element in List");
+        }
+        return oldValue;
```

Comment:
in this method you need to find an index of the element and call the method `remove(int index)`, otherwise - throw an exception

Reply:
In this method I've found the the index and called the method remove(int index)